src/core/credentials/transport/tls/spiffe_utils.cc

```diff
@@ -0,0 +1,158 @@
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_split.h"
+
+constexpr absl::string_view SPIFFE_PREFIX = "spiffe://";
+
+namespace grpc_core {
+namespace experimental {
+namespace {
+
+absl::Status doInitialUriValidation(absl::string_view uri) {
+  if (uri.empty()) {
+    return absl::InvalidArgumentError(
+        "SpiffeId cannot be parsed from empty uri");
+  }
+  if (uri.length() > 2048) {
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "URI length is %d, maximum allowed for SPIFFE ID is 2048",
+        uri.length()));
+  }
+  if (absl::StrContains(uri, "#")) {
+    return absl::InvalidArgumentError(
+        "SPIFFE ID cannot contain query fragments");
+  }
+  if (absl::StrContains(uri, "?")) {
+    return absl::InvalidArgumentError(
+        "SPIFFE ID cannot contain query parameters");
+  }
+  return absl::OkStatus();
+}
+
+absl::Status validateTrustDomain(absl::string_view trust_domain) {
+  if (trust_domain.empty()) {
+    return absl::InvalidArgumentError("Trust domain cannot be empty");
+  }
+  if (trust_domain.size() >= 256) {
+    return absl::InvalidArgumentError(
+        "Trust domain maximum length is 255 characters");
+  }
+  for (auto c : trust_domain) {
+    if (c >= 'a' && c <= 'z') continue;
+    if (c >= '0' && c <= '9') continue;
+    if (c == '.') continue;
+    if (c == '-') continue;
+    if (c == '_') continue;
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "Trust domain contains invalid character %c. MUST contain only "
+        "lowercase letters, numbers, dots, dashes, and underscores",
+        c));
+  }
+  return absl::OkStatus();
+}
+
+absl::Status validatePathSegment(absl::string_view path_segment) {
+  if (path_segment.empty()) {
+    return absl::InvalidArgumentError("Path segment cannot be empty");
+  }
+  if (path_segment == "." || path_segment == "..") {
+    return absl::InvalidArgumentError(
+        "Path segment cannot be a relative modifier (. or ..)");
+  }
+  for (auto c : path_segment) {
+    if (c >= 'a' && c <= 'z') continue;
+    if (c >= 'A' && c <= 'Z') continue;
+    if (c >= '0' && c <= '9') continue;
+    if (c == '.') continue;
+    if (c == '-') continue;
+    if (c == '_') continue;
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "Path segment contains invalid character %c. MUST contain only "
+        "letters, numbers, dots, dashes, and underscores",
+        c));
+  }
+  return absl::OkStatus();
+}
+
+absl::Status validatePath(absl::string_view path) {
+  if (path.empty()) {
+    return absl::OkStatus();
+  }
+  for (absl::string_view segment : absl::StrSplit(path, '/')) {
+    if (absl::Status status = validatePathSegment(segment); !status.ok()) {
+      return status;
+    }
+  }
+  return absl::OkStatus();
+}
+
+}  // namespace
+
+// spiffe://trust-domain-name/path
+absl::StatusOr<SpiffeId> SpiffeId::FromString(absl::string_view uri) {
+  if (absl::Status status = doInitialUriValidation(uri); !status.ok()) {
```

Comment:
It looks like gRPC now has the `GRPC_RETURN_IF_ERROR` macro. Could we use that here and throughout the class?

Reply:
Done